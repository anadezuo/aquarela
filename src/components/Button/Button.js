import React from "react";
import propTypes from "prop-types";
import styled from "styled-components";
import BUTTON_COLORS from "../../constants/ButtonColors";
import BUTTON_VARIANTS from "../../constants/ButtonVariants";

const getMainColor = ({ theme, color }) => {
  return {
    [BUTTON_COLORS.DEFAULT]: "#e0e0e0",
    [BUTTON_COLORS.PRIMARY]: theme.colors.primary.main,
    [BUTTON_COLORS.DANGER]: theme.colors.danger.main,
  }[color];
};

const getDarkColor = ({ theme, color }) => {
  return {
    [BUTTON_COLORS.DEFAULT]: "#5a6268",
    [BUTTON_COLORS.PRIMARY]: theme.colors.primary.dark,
    [BUTTON_COLORS.DANGER]: theme.colors.danger.dark,
  }[color];
};

const getOutlinedText = (props) => {
  if (props.color === BUTTON_COLORS.DEFAULT) {
    return "#212121";
  }

  return getMainColor(props);
};

const getLinkText = (props) => {
  if (props.color === BUTTON_COLORS.DEFAULT) {
    return "#757575";
  }

  return getMainColor(props);
};

const ButtonBase = styled.button`
  font-size: 1 rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  background-color: ${getMainColor};
  border: 2px solid ${getMainColor};
  color: ${({ theme }) => theme.colors.primary.text};

  &:disabled {
    opacity: 0.6;
    cursor: not-allowed;
  }

  &:hover:enabled {
    background-color: ${getDarkColor};
    border: 2px solid ${getDarkColor};
  }
`;

const ButtonOutlined = styled(ButtonBase)`
  background-color: transparent;
  color: ${getOutlinedText};

  &:hover:enabled {
    background-color: transparent;
    color: ${getDarkColor};
  }
`;

const ButtonLink = styled(ButtonBase)`
  background-color: transparent;
  border-color: transparent;
  color: ${getLinkText};
  padding-left: 0;
  padding-right: 0;
`;

const Button = (props) => {
  switch (props.variant) {
    case BUTTON_VARIANTS.OUTLINED:
      return <ButtonOutlined {...props} />;
    case BUTTON_VARIANTS.LINK:
      return <ButtonLink {...props} />;
    default:
      return <ButtonBase {...props} />;
  }
};

Button.propTypes = {
  type: propTypes.string,
  children: propTypes.node,
  color: propTypes.oneOf(Object.values(BUTTON_COLORS)),
  variant: propTypes.oneOf(Object.values(BUTTON_VARIANTS)),
};

Button.defaultProps = {
  type: "button",
  children: undefined,
  color: BUTTON_COLORS.DEFAULT,
  variant: BUTTON_VARIANTS.DEFAULT,
};

export default Button;
