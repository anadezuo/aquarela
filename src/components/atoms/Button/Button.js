import React from "react";
import propTypes from "prop-types";
import styled from "styled-components";
import BUTTON_COLORS from "constants/ButtonColors";
import BUTTON_VARIANTS from "constants/ButtonVariants";

const getMainColor = ({ theme, color }) => {
  return {
    [BUTTON_COLORS.DEFAULT]: "#e0e0e0",
    [BUTTON_COLORS.PRIMARY]: theme.colors.primary.main,
    [BUTTON_COLORS.DANGER]: theme.colors.danger.main,
  }[color];
};

const getDarkColor = ({ theme, color }) => {
  return {
    [BUTTON_COLORS.DEFAULT]: "#5a6268",
    [BUTTON_COLORS.PRIMARY]: theme.colors.primary.dark,
    [BUTTON_COLORS.DANGER]: theme.colors.danger.dark,
  }[color];
};

const getTextColor = ({ theme, color }) => {
  return {
    [BUTTON_COLORS.DEFAULT]: "#212121",
    [BUTTON_COLORS.PRIMARY]: theme.colors.primary.text,
    [BUTTON_COLORS.DANGER]: theme.colors.danger.text,
  }[color];
};

const getOutlinedText = (props) => {
  if (props.color === BUTTON_COLORS.DEFAULT) {
    return "#212121";
  }

  return getMainColor(props);
};

const getLinkText = (props) => {
  if (props.color === BUTTON_COLORS.DEFAULT) {
    return "#757575";
  }

  return getMainColor(props);
};

const ButtonBase = styled.button`
  font-size: 1 rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 8px 30px;
  background-color: ${getMainColor};
  border: 2px solid ${getMainColor};
  color: ${getTextColor};
  display: inline-block;
  text-decoration: none;

  &:disabled {
    opacity: 0.6;
    cursor: not-allowed;
  }

  &:hover:not(:disabled) {
    background-color: ${getDarkColor};
    border: 2px solid ${getDarkColor};
  }
`;

const ButtonOutlined = styled(ButtonBase)`
  background-color: transparent;
  color: ${getOutlinedText};

  &:hover:not(:disabled) {
    background-color: transparent;
    color: ${getDarkColor};
  }
`;

const ButtonLink = styled(ButtonBase)`
  background-color: transparent;
  border-color: transparent;
  color: ${getLinkText};
  padding-left: 0;
  padding-right: 0;
  &:hover:not(:disabled) {
    background-color: transparent;
    border-color: transparent;
    color: ${getDarkColor};
  }
`;

const Button = (props) => {
  switch (props.variant) {
    case BUTTON_VARIANTS.OUTLINED:
      return <ButtonOutlined {...props} />;
    case BUTTON_VARIANTS.LINK:
      return <ButtonLink {...props} />;
    default:
      return <ButtonBase {...props} />;
  }
};

Button.propTypes = {
  type: propTypes.string,
  children: propTypes.node,
  color: propTypes.oneOf(Object.values(BUTTON_COLORS)),
  variant: propTypes.oneOf(Object.values(BUTTON_VARIANTS)),
};

Button.defaultProps = {
  type: "button",
  children: undefined,
  color: BUTTON_COLORS.DEFAULT,
  variant: BUTTON_VARIANTS.DEFAULT,
};

export default Button;
