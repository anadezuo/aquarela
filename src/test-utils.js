import React from "react";
import ThemeProvider from "styles/ThemeProvider";
import { render } from "@testing-library/react";

const TestProvider = ({ children }) => (
  <ThemeProvider>{children}</ThemeProvider>
);

export * from "@testing-library/react";
export { default as userEvent } from "@testing-library/user-event";

const customRender = (ui, options) =>
  render(ui, { wrapper: TestProvider, ...options });

export { customRender as render };
