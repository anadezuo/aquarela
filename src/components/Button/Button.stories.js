import React from "react";
import styled from "styled-components";
import { actions } from "@storybook/addon-actions";
import BUTTON_COLORS from "../../constants/ButtonColors";
import Button from "./Button";
import BUTTON_VARIANTS from "constants/ButtonVariants";

export default {
  title: "Components/Button",
  component: Button,
};

const Toolbar = styled.div`
  & > * {
    margin-right: 8px;
  }
`;

const events = actions({ onClick: "clicked" });

export const usage = () => (
  <div>
    <p>Enabled:</p>
    <Toolbar>
      <Button {...events}>Default</Button>
      <Button color={BUTTON_COLORS.PRIMARY} {...events}>
        Primary
      </Button>
      <Button color={BUTTON_COLORS.DANGER} {...events}>
        Danger
      </Button>
    </Toolbar>
    <p>Disabled:</p>
    <Toolbar>
      <Button disabled {...events}>
        Default
      </Button>
      <Button color={BUTTON_COLORS.PRIMARY} disabled {...events}>
        Primary
      </Button>
      <Button color={BUTTON_COLORS.DANGER} disabled {...events}>
        Danger
      </Button>
    </Toolbar>
  </div>
);

export const Outlined = () => (
  <div>
    <p>Enabled:</p>
    <Toolbar>
      <Button variant={BUTTON_VARIANTS.OUTLINED} {...events}>
        Default
      </Button>

      <Button
        color={BUTTON_COLORS.PRIMARY}
        variant={BUTTON_VARIANTS.OUTLINED}
        {...events}
      >
        Primary
      </Button>
      <Button
        color={BUTTON_COLORS.DANGER}
        variant={BUTTON_VARIANTS.OUTLINED}
        {...events}
      >
        Danger
      </Button>
    </Toolbar>
    <p>Disabled:</p>
    <Toolbar>
      <Button variant={BUTTON_VARIANTS.OUTLINED} disabled {...events}>
        Default
      </Button>
      <Button
        color={BUTTON_COLORS.PRIMARY}
        variant={BUTTON_VARIANTS.OUTLINED}
        disabled
        {...events}
      >
        Primary
      </Button>
      <Button
        color={BUTTON_COLORS.DANGER}
        variant={BUTTON_VARIANTS.OUTLINED}
        disabled
        {...events}
      >
        Danger
      </Button>
    </Toolbar>
  </div>
);

export const Link = () => (
  <div>
    <p>Enabled:</p>
    <Toolbar>
      <Button variant={BUTTON_VARIANTS.LINK} {...events}>
        Default
      </Button>

      <Button
        color={BUTTON_COLORS.PRIMARY}
        variant={BUTTON_VARIANTS.LINK}
        {...events}
      >
        Primary
      </Button>
      <Button
        color={BUTTON_COLORS.DANGER}
        variant={BUTTON_VARIANTS.LINK}
        {...events}
      >
        Danger
      </Button>
    </Toolbar>
    <p>Disabled:</p>
    <Toolbar>
      <Button variant={BUTTON_VARIANTS.LINK} disabled {...events}>
        Default
      </Button>
      <Button
        color={BUTTON_COLORS.PRIMARY}
        variant={BUTTON_VARIANTS.LINK}
        disabled
        {...events}
      >
        Primary
      </Button>
      <Button
        color={BUTTON_COLORS.DANGER}
        variant={BUTTON_VARIANTS.LINK}
        disabled
        {...events}
      >
        Danger
      </Button>
    </Toolbar>
  </div>
);
