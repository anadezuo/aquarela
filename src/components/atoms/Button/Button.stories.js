import React from "react";
import styled from "styled-components";
import { actions } from "@storybook/addon-actions";
import BUTTON_COLORS from "constants/ButtonColors";
import Button from "./Button";
import BUTTON_VARIANTS from "constants/ButtonVariants";

export default {
  title: "Components/Atoms/Button",
  component: Button,
};

const Toolbar = styled.div`
  & > * {
    margin-right: 8px;
  }
`;

const ButtonDefault = ({ ...rest }) => (
  <Button {...events} {...rest}>
    Default
  </Button>
);

const ButtonPrimary = ({ ...rest }) => (
  <Button color={BUTTON_COLORS.PRIMARY} {...events} {...rest}>
    Primary
  </Button>
);

const ButtonDanger = ({ ...rest }) => (
  <Button color={BUTTON_COLORS.DANGER} {...events} {...rest}>
    Primary
  </Button>
);

const ButtonsWrapper = ({ ...rest }) => (
  <div>
    <p>Enabled:</p>
    <Toolbar>
      <ButtonDefault {...events} {...rest} />
      <ButtonPrimary {...events} {...rest} />
      <ButtonDanger {...events} {...rest} />
    </Toolbar>
    <p>As a link:</p>
    <Toolbar>
      <ButtonDefault as="a" href="#" {...events} {...rest} />
      <ButtonPrimary as="a" href="#" {...events} {...rest} />
      <ButtonDanger as="a" href="#" {...events} {...rest} />
    </Toolbar>
    <p>Disabled:</p>
    <Toolbar>
      <ButtonDefault disabled {...events} {...rest} />
      <ButtonPrimary disabled {...events} {...rest} />
      <ButtonDanger disabled {...events} {...rest} />
    </Toolbar>
  </div>
);

const events = actions({ onClick: "clicked" });

export const usage = () => <ButtonsWrapper />;

export const Outlined = () => (
  <ButtonsWrapper variant={BUTTON_VARIANTS.OUTLINED} />
);

export const Link = () => <ButtonsWrapper variant={BUTTON_VARIANTS.LINK} />;
