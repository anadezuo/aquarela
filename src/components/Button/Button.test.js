import { fireEvent } from "@testing-library/react";
import { render } from "test-utils";
import BUTTON_COLORS from "../../constants/ButtonColors";
import BUTTON_VARIANTS from "../../constants/ButtonVariants";
import Button from "./Button";

describe("Button", () => {
  function mount({ children, color, variant, disabled, onClick }) {
    return render(
      <Button
        children={children}
        color={color}
        variant={variant}
        disabled={disabled}
        onClick={onClick}
      />
    );
  }

  function getButton(getByRole) {
    return getByRole("button");
  }

  it("should render a text", () => {
    const children = "My button";
    const { getByText } = mount({ children });
    expect(getByText(children)).toBeInTheDocument();
  });

  it("should render disabled when value is given", () => {
    const { getByRole } = mount({ disabled: true });
    const button = getButton(getByRole);
    expect(button).toHaveAttribute("disabled", "");
  });

  it("should execute listener callback on click", () => {
    const onClick = jest.fn();
    const { getByRole } = mount({ onClick });
    fireEvent.click(getByRole("button"));
    expect(onClick).toBeCalled();
  });

  it.each(Object.values(BUTTON_COLORS).map((item) => [item]))(
    "should show %s color when color is provided",
    (color) => {
      const { asFragment } = mount({ color });
      expect(asFragment()).toMatchSnapshot();
    }
  );

  it.each(Object.values(BUTTON_VARIANTS).map((item) => [item]))(
    "should optionally render an %s button",
    (variant) => {
      const { asFragment } = mount({ variant });
      expect(asFragment()).toMatchSnapshot();
    }
  );

  it("should optionally render an outlined button with color primary", () => {
    const { asFragment } = mount({
      color: BUTTON_COLORS.PRIMARY,
      variant: BUTTON_VARIANTS.OUTLINED,
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it("should optionally render an link button with color primary", () => {
    const { asFragment } = mount({
      color: BUTTON_COLORS.PRIMARY,
      variant: BUTTON_VARIANTS.LINK,
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
