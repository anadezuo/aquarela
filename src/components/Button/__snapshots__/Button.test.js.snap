// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should optionally render an default button 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 1 rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  background-color: #e0e0e0;
  border: 2px solid #e0e0e0;
  color: #212121;
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

.c0:hover:enabled {
  background-color: #5a6268;
  border: 2px solid #5a6268;
}

<button
    class="c0"
    color="default"
    type="button"
  />
</DocumentFragment>
`;

exports[`Button should optionally render an link button 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 1 rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  background-color: #e0e0e0;
  border: 2px solid #e0e0e0;
  color: #212121;
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

.c0:hover:enabled {
  background-color: #5a6268;
  border: 2px solid #5a6268;
}

.c1 {
  background-color: transparent;
  border-color: transparent;
  color: #757575;
  padding-left: 0;
  padding-right: 0;
}

<button
    class="c0 c1"
    color="default"
    type="button"
  />
</DocumentFragment>
`;

exports[`Button should optionally render an link button with color primary 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 1 rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  background-color: #ffc107;
  border: 2px solid #ffc107;
  color: #212121;
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

.c0:hover:enabled {
  background-color: #c79100;
  border: 2px solid #c79100;
}

.c1 {
  background-color: transparent;
  border-color: transparent;
  color: #ffc107;
  padding-left: 0;
  padding-right: 0;
}

<button
    class="c0 c1"
    color="primary"
    type="button"
  />
</DocumentFragment>
`;

exports[`Button should optionally render an outlined button 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 1 rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  background-color: #e0e0e0;
  border: 2px solid #e0e0e0;
  color: #212121;
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

.c0:hover:enabled {
  background-color: #5a6268;
  border: 2px solid #5a6268;
}

.c1 {
  background-color: transparent;
  color: #212121;
}

.c1:hover:enabled {
  background-color: transparent;
  color: #5a6268;
}

<button
    class="c0 c1"
    color="default"
    type="button"
  />
</DocumentFragment>
`;

exports[`Button should optionally render an outlined button with color primary 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 1 rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  background-color: #ffc107;
  border: 2px solid #ffc107;
  color: #212121;
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

.c0:hover:enabled {
  background-color: #c79100;
  border: 2px solid #c79100;
}

.c1 {
  background-color: transparent;
  color: #ffc107;
}

.c1:hover:enabled {
  background-color: transparent;
  color: #c79100;
}

<button
    class="c0 c1"
    color="primary"
    type="button"
  />
</DocumentFragment>
`;

exports[`Button should show danger color when color is provided 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 1 rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  background-color: #ff3d00;
  border: 2px solid #ff3d00;
  color: #212121;
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

.c0:hover:enabled {
  background-color: #b22a00;
  border: 2px solid #b22a00;
}

<button
    class="c0"
    color="danger"
    type="button"
  />
</DocumentFragment>
`;

exports[`Button should show default color when color is provided 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 1 rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  background-color: #e0e0e0;
  border: 2px solid #e0e0e0;
  color: #212121;
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

.c0:hover:enabled {
  background-color: #5a6268;
  border: 2px solid #5a6268;
}

<button
    class="c0"
    color="default"
    type="button"
  />
</DocumentFragment>
`;

exports[`Button should show primary color when color is provided 1`] = `
<DocumentFragment>
  .c0 {
  font-size: 1 rem;
  font-weight: 600;
  text-transform: uppercase;
  padding: 12px 36px;
  background-color: #ffc107;
  border: 2px solid #ffc107;
  color: #212121;
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
}

.c0:hover:enabled {
  background-color: #c79100;
  border: 2px solid #c79100;
}

<button
    class="c0"
    color="primary"
    type="button"
  />
</DocumentFragment>
`;
